cmd/cronosd/cmd/reindex-duplicated-tx.go

```diff
@@ -0,0 +1,205 @@
+	sm "github.com/tendermint/tendermint/state"
+	"github.com/tendermint/tendermint/state/indexer/sink/psql"
+	"github.com/tendermint/tendermint/state/txindex"
+	"github.com/tendermint/tendermint/state/txindex/kv"
+	"github.com/tendermint/tendermint/state/txindex/null"
+	tmstore "github.com/tendermint/tendermint/store"
+)
+
+const FlagConcurrency = "concurrency"
+
+// ReindexDuplicatedTx update the tx execution result of false-failed tx in tendermint db
+func ReindexDuplicatedTx() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "reindex-duplicated-tx",
+		Short: "Reindex tx that suffer from tendermint duplicated tx issue",
+		Args:  cobra.ExactArgs(2),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			ctx := server.GetServerContextFromCmd(cmd)
+
+			chainID, err := cmd.Flags().GetString(flags.FlagChainID)
+			if err != nil {
+				return err
+			}
+			startHeight, err := strconv.ParseInt(args[0], 10, 64)
+			if err != nil {
+				return err
+			}
+			endHeight, err := strconv.ParseInt(args[1], 10, 64)
+			if err != nil {
+				return err
+			}
+			if startHeight < 1 {
+				return fmt.Errorf("invalid start-block: %d", startHeight)
+			}
+			if endHeight < startHeight {
+				return fmt.Errorf("invalid end-block %d, smaller than start-block", endHeight)
+			}
+
+			tmDB, err := openTMDB(ctx.Config, chainID)
+			if err != nil {
+				return err
+			}
+
+			// iterate and patch a single block
+			processBlock := func(height int64) error {
+				blockResult, err := tmDB.stateStore.LoadABCIResponses(height)
+				if err != nil {
+					return err
+				}
+				block := tmDB.blockStore.LoadBlock(height)
+				if err != nil {
+					return err
+				}
+				if block == nil {
+					return fmt.Errorf("block not found: %d", height)
+				}
+
+				for txIndex, txResult := range blockResult.DeliverTxs {
+					tx := block.Txs[txIndex]
+					txHash := tx.Hash()
+					indexed, err := tmDB.txIndexer.Get(txHash)
+					if err != nil {
+						return err
+					}
+					if txResult.Code == 0 && txResult.Code != indexed.Result.Code {
+						// a success tx but indexed wrong, reindex the tx
+						result := &abci.TxResult{
+							Height: height,
+							Index:  uint32(txIndex),
+							Tx:     tx,
+							Result: *txResult,
+						}
+
+						if err := tmDB.txIndexer.Index(result); err != nil {
+							return err
+						}
+					}
+				}
+
+				return nil
+			}
+
+			concurrency, err := cmd.Flags().GetInt(FlagConcurrency)
+			if err != nil {
+				return err
+			}
+
+			blockChan := make(chan int64, concurrency)
+			var wg sync.WaitGroup
+			ctCtx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			for i := 0; i < concurrency; i++ {
+				wg.Add(1)
+				go func(wg *sync.WaitGroup) {
+					defer wg.Done()
+					for {
+						select {
+						case <-ctCtx.Done():
+							return
+						case blockNum, ok := <-blockChan:
+							if !ok {
+								return
+							}
+
+							if err := processBlock(blockNum); err != nil {
+								cancel()
```

Comment:
it seems we can not distinguish processBlock failure from normal final cancellations

Reply:
yeah, any suggestions?

Reply:
I guess the cancel will be revealed in `ctCtx.Err()`?
probably should log the err here, for trouble shooting purpose.

Reply:
added a log.